x-pack/plugins/fleet/server/services/setup.ts

```diff
@@ -334,3 +342,68 @@ export async function ensureFleetDirectories() {
     );
   }
 }
+
+/**
+ * Find preexisting fleet-managed data views like `logs-*` and `metrics-*` and update their `title` and `name` attributes
+ * as needed based on the newer values. See https://github.com/elastic/kibana/issues/120340.
+ */
+export async function migrateLegacyDataViews({
+  savedObjectsClient,
+  logger,
+}: {
+  savedObjectsClient: SavedObjectsClientContract;
+  logger: Logger;
+}) {
+  const indexPatternSavedObjects = getFleetManagedDataViewDefinitions();
+
+  const existingDataViews = [];
+
+  // TODO - Might be worth refactoring this to a `bulkGet` or `find` instead, though Fleet only manages
+  // two dataviews right now so the gains are currently negligible.
+  for (const dataView of indexPatternSavedObjects) {
+    try {
+      const existingDataView = await savedObjectsClient.get<DataViewSavedObjectAttrs>(
+        KibanaSavedObjectType.indexPattern,
+        dataView.id,
+        {}
+      );
+
+      existingDataViews.push(existingDataView);
+    } catch (error) {
+      // If an existing data view is not found, ignore the erorr
+      if (!SavedObjectsErrorHelpers.isNotFoundError(error)) {
+        throw error;
+      }
+    }
+  }
+
+  for (const existingDataView of existingDataViews) {
+    const matchingDataView = indexPatternSavedObjects.find(({ id }) => id === existingDataView.id)!;
+
+    // If a data view with the given ID already exists, we need to make sure it has the right `title` and `name`.
+    const shouldUpdateTitle =
+      existingDataView.attributes.title !== matchingDataView.attributes.title;
+    const shouldUpdateName = existingDataView.attributes.name !== matchingDataView.attributes.name;
+
+    let updateAttributes: Partial<DataViewSavedObjectAttrs> = {};
+
+    if (shouldUpdateTitle) {
+      updateAttributes = { ...updateAttributes, title: matchingDataView.attributes.title };
+    }
+    if (shouldUpdateName) {
+      updateAttributes = { ...updateAttributes, name: matchingDataView.attributes.name };
+    }
+
+    if (!isEmpty(updateAttributes)) {
+      logger.info(
+        `Found outdated data view with id: ${matchingDataView.id}. Updating name to "${updateAttributes.name}" and title to "${updateAttributes.title}".`
+      );
+
+      await savedObjectsClient.update(
```

Comment:
could we have a conflict if multiple kibana do the upgrade at the same time? and should we handle it

Reply:
I think it is technically possible to have a conflict in this case, but because setup is called on every page load we should have eventual consistency here. 

Still, I think we could probably just retry on conflict here for safety.